### ISSUE TITLE
Release Engineering/Minimizing Friction Around Releases

### ISSUE THREAD

**Issue by bashtage** [statsmodels/statsmodels#4058]
There are a number of improvements that could really simplify releasing statsmodels.  Implementing these are different from writing/testing statistical functions.

The following changes would help improve the setup/release of statsmodels

- [x] Use doctr to push doc builds to /devel on every successful commit
- [x] Use doctr to push doc builds to /stable (or just /) on every tagged commit
- [x] Switch to versioneer and remove all explicit version string setting -- the setting a tag would be the one and only place that a version would be updated #4035
- [x] Fix setup so that it can run in an empty environment #3207  #3195
- [x] Ensure requirements for build and docs are correct #3194 
- [x] Switch clean to "Use git clean -Xfd" (like NumPy). See https://github.com/numpy/numpy/blob/4e6a8f0efa09f3af04b900c1ec796622eaf8141c/setup.py#L301 
- [ ] Build nightly windows wheels and/or conda builds and push on successful master builds.  It is also possible to capture wheels on tags so that these can be posted to PyPi so simplify releasing. #3786 
- [x] Simplify setup to that it uses less homegrown code and relies more on upstream projects.  For example use Cython to determine caching.  Partially in #4035 . Pandas has implemented this I think.
- [x] Enable building with coverage when requested #4035 

Some of @TomAugspurger @thequackdaddy @jbrockmendel @matthew-brett @rgommers might have a lot of experience in these.

**Comment by jbrockmendel** [comment]
This sounds like a good idea.  I'll volunteer some labor.

**Comment by matthew-brett** [comment]
It's pretty easy to set up daily wheel builds - we already do this for numpy, scipy, scikit-image and others.

I personally like and use Versioneer - but it can be a bit fiddly - pip installs can mess it up because pip copies all the files out to a directory, at which point you've lost the git information you need - you can work round that, but you have to be aware.  Also, if you want to patch the source before building, it will change the version to something `-dirty`.  Maybe minor problems, but worth being aware of them.

I wouldn't recommend uploading directly to pypi - it's just too accident-prone - its easy to run into unexpected failures and then have a half-baked release.  The other packages have an intermediate step where you wait for the wheels to build and test and upload to Rackspace, and then you run a script to upload the wheels to pypi from Rackspace.  It's one more step, but gives an extra manual check that everything is OK before doing a release.

**Comment by bashtage** [comment]
Yes, I wasn't envisioning directly uploading to PyPi.  It is pretty annoying to upload the wrong wheel (although taggs sort of help). 

I've experienced the dirty issue.  I suppose one has to be careful to know when they are really ready to release.  One challenge is maybe when there are files that need to go into the src dist that aren't in the repo, although I think this is OK since they aren't tracked. 

I think have the following steps automatic would help:

One release tag:
1. Automatic build/push docs 
2. Windows wheels built and ready for upload to PyPi

One a master commit:
1. Windows wheels/conda packages built and uploaded to somewhere doscoverable. 
2. Dev docs built/pushed

This is quite different from current release path, so @josef-pkt will need to be onboard.

**Comment by matthew-brett** [comment]
I am sure you also meant macOS and manylinux wheels ...  They already get built by default : https://travis-ci.org/MacPython/statsmodels-wheel

**Comment by bashtage** [comment]
I suppose so.  If it is free, why not 👍 . I can see the case for OSX, but the case for Linux is less compelling for SM than for more complicated builds (MPL, NumPy, SciPy).

**Comment by matthew-brett** [comment]
I think the point for manylinux wheels is that, because virtually every other useful package does them (numpy, scipy, matplotlib, pandas, h5py, scikit-image, scikit-learn ...), you probably don't want statsmodels to be the one that breaks without a compiler.

**Comment by jbrockmendel** [comment]
Given that the rest of the ecosystem is doing this, is there any config/infrastructure/institutional know-how that statsmodels can piggy-back on?

**Comment by matthew-brett** [comment]
For the wheels - yes - it's already done, bar some small mods - over at https://github.com/MacPython/statsmodels-wheels .  That should build the wheels for release.  I can also hook up daily wheel builds, that just takes a few lines.

**Comment by bashtage** [comment]
I think we should wait on daily wheels since these are really developer/testing only.  

Call this issue all done.
